### DISCUSSION TITLE
How to use the new validation backend introduced in V3?

### DISCUSSION THREAD

**Discussion by francoism90** [spatie/laravel-data#347] [Q&A]
> rules are now generated based upon the payload provided, not what a payload possibly could be. This means rules can change depending on the provided payload.

This is the user dataobject:
```php
class UserData extends Data implements Wireable
{
    use WireableData;

    public function __construct(
        public Optional|string $id,
        public Optional|Lazy|string $email,
        public Optional|Lazy|null|string $password,
        public Optional|Lazy|null|string $firstname,
        public Optional|Lazy|null|string $lastname,
    ) {
    }

    public static function rules(array $payload = []): array
    {
        return [
            'id' => ['required', 'string', new UserExists()],
            'email' => ['required', 'email', Rule::unique('users', 'email')->ignore($payload['id'] ?? null, 'id')],
            'password' => ['required', 'string', Password::default()],
            'firstname' => ['required', 'string', 'max:20'],
            'lastname' => ['required', 'string', 'max:40'],
        ];
    }
```
To fill the rules payload, I need to inject the `rules` like this:
```
protected function rules(): array
    {
        return Arr::prependKeysWith(
            array: UserData::rules($this->user->toArray()), // e.g. { id: .., firstname: 'foo', email: 'email@example.com' }
            prependWith: 'user.'
        );
    }
```
On every (form) submit, it will use the Livewire's dataobject to populate the `$payload`. I don't know any other workarounds, so please let me know if they are.

Questions:
- It only works when prepending `sometimes` to the rules, to exclude non existing data entries (e.g. `password`). Do I need to use a different set of rules?
- Do you still need to use `Optional`?
- Any tips for Livewire?

Many thanks!

**Comment by rubenvanassche** [comment]
I haven't used LiveWire that much so I'm afraid I cannot help you with this.

Since you're working with a root data payload, all validation rules will always be generated for all properties of the object. Only when nesting other data objects which are optional then we'll stop generating rules based upon the given payload. So my guess is that Optional will still be required in your case.

As a rule of thumb, though the automatic generation of validation rules of this package is really nice. It is sometimes a lot easier in cases like this one to just manually write them out 😄

**Reply by francoism90** [reply]
@rubenvanassche I've ended up using this:
```
public UserData $user;

protected function userRules(): array
    {
        return Arr::prependKeysWith(
            array: UserData::rules($this->userValidationContext()),
            prependWith: 'user.'
        );
    }

protected function userValidationContext(): ValidationContext
    {
        return new ValidationContext(
            payload: $this->user->toArray(), // is UserData object variable
            fullPayload: $this->all(), // returns all public properties from Livewire
            path: new ValidationPath(null), // seems this is something that can be customised?
        );
    }
```
It seems Livewire call's the rules when doing `$this->validate()`, so we get 'fresh' data each time.
